docs/ide/api-designs/Workspace and Source Generated Documents.md

```diff
@@ -0,0 +1,73 @@
+
+The `Id` of the `Document` will be generated dynamically by the workspace. As much as possible, the Id will be persistent between snapshots as much as is practical. When a generator produces a document, it provides a "hint name" which is intended to be used for the IDE to allow for 'permanence' between runs. `DocumentId`s are defined by a GUID in the implementation, so to ensure the DocumentId can be stable, we will generate the GUID as a hash of the hint name, the generator name, and any other peritnent information. Although the implementation may do extra caching behind the scenes where it won't have to recompute hashes in this way, this will ensure the `DocumentId` stays stable even if the caches fail to avoid any surprises.
+
+`SourceGeneratedDocument` will have some additional members:
+
+1. `HintName`: the exact hint name that was provided by the generator.
+1. `SourceGenerator`: the ISourceGenerator instance that produced this document.
+
+For now, any _mutation_ of a SourceGeneratedDocument will not be supported. Calls to the following members will throw `NotSupportedException`:
+
+1. `WithText`
+2. `WithSyntaxRoot`
+3. `WithName`
+4. `WithFolders`
+5. `WithFilePath`
+
+Down the road, I hope that we might offer an extension point for a source generator to participate in WithText and WithSyntaxRoot. In many cases the generated source may contain expressions that are verbatim C# (think a .cshtml file having a C# section), and so we could have WithText or WithSyntaxRoot apply the resulting edit back to the document that the verbatim C# came from. The design of this is out of scope here.
+
+## APIs on `Project`
+
+### Existing API: `Documents`
+
+Note the Documents collection will _not_ change, and will only include regular documents. It cannot include source generated documents because since it is a property access, there is no opportunity to make this async, nor have any way to express cancellation. Since a generator could produce any number of documents, there's simply no way to make this answer quickly without fundamentally taking a different approach to the API.
+
+### New API: `ImmutableArray<SourceGeneratedDocument> GetSourceGeneratedDocumentsAsync(CancelationToken)`
+
+This will run generators if they have not already ran, and then return a `SourceGeneratedDocument` for each generated document.
+
+The implementation of this will run `GetCompilationAsync` if the documents have not been generated; if the Compilation had already been generated, we would have cached the generated document information and so this would be cheap. We will hold onto the list of documents strongly (the tree will be a recoverable tree), so even if the `Compilation` is GC'ed we won't have to recompute this part a second time.
+
+### New API: `SourceGeneratedDocument? GetSourceGeneratedDocumentAsync(DocumentId, CancellationToken)`
+
+Fetches a single document by ID; equivalent to calling the API and filtering down to the right document.
```

Comment:
Is there scope here to _just_ run the generator that's needed to get the document?

Reply:
Right now this is mostly a convenience method. It'll also have an advantage going forward because if you're operating in open source generated file, we can ensure this fast-paths to return the generated text already being displayed as opposed to having to compute other stuff that doesn't matter.